dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/DockerRegistryAPIHelper.java

```diff
@@ -0,0 +1,236 @@
+                return Optional.empty();
+            }
+        } catch (IOException | InterruptedException ex) {
+            LOG.error("Could not send manifest request GET {}", getManifestURL, ex);
+            return Optional.empty();
+        }
+
+        return Optional.of(manifestResponse);
+    }
+
+    /*
+     * Returns the next wait interval, in milliseconds, using an exponential
+     * backoff algorithm.
+     */
+    @SuppressWarnings("checkstyle:MagicNumber")
+    public static long getWaitTimeExp(int retryCount) {
+        if (0 == retryCount) {
+            return 0;
+        }
+
+        long waitTime = ((long) Math.pow(2, retryCount) * 100L);
+
+        return waitTime;
+    }
+
+    /**
+     * Get a blob specified by digest for a Docker image by calling the Docker Registry HTTP API V2 endpoint: GET /v2/[repo]/blobs/[digest]
+     *
+     * @param token Authentication token with pull access for the image's repository
+     * @param registryDockerPath
+     * @param repo
+     * @param digest SHA256 digest of the blob to download
+     * @return HttpResponse
+     */
+    public Optional<Response> getDockerBlob(String token, String registryDockerPath, String repo, String digest) {
+        // Ex: https://ghcr.io/v2/<repo>/blobs/<digest>
+        String getBlobURL = String.format("https://%s/v2/%s/blobs/%s", registryDockerPath, repo, digest);
+
+        Request request = new Request.Builder()
+                .url(getBlobURL)
+                .addHeader(HttpHeaders.AUTHORIZATION, String.join(" ", JWT_SECURITY_DEFINITION_NAME, token))
+                .build();
+
+        Response blobResponse;
+        try {
+            // This endpoint may issue a 307 redirect to another service to download the blob
+            blobResponse = client.newCall(request).execute();
+        } catch (IOException ex) {
+            LOG.error("Could not send blob request GET {}", getBlobURL, ex);
+            return Optional.empty();
+        }
+
+        if (blobResponse.isSuccessful()) {
+            return Optional.of(blobResponse);
+        } else {
+            Optional<Map<String, String>> error = getDockerError(blobResponse);
+            if (error.isPresent()) {
+                LOG.error("Blob response has error code {} '{}' with message '{}'", blobResponse.code(), error.get().get("code"), error.get().get("message"));
+            }
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Gets the error map of an unsuccessful Docker Registry HTTP API V2 response
+     * Failures are reported as part of 4xx responses, in a json response body (https://docs.docker.com/registry/spec/api/#errors)
+     *
+     * @param response Docker Registry HTTP API V2 response
+     * @return an error map of the response if it exists
+     */
+    public Optional<Map<String, String>> getDockerError(Response response) {
+        Map<String, List<Map<String, String>>> errorMap = GSON.fromJson(response.body().charStream(), Map.class);
```

Comment:
I think we discussed this in a previous PR, so you may have already answered this, what happens if it's a 5xx response, e.g., their service is down? You may not even get a response body, or it may not be JSON.

Reply:
I'm not really sure what happens if it's a 5xx response, so I changed the method to check if it's a 4xx response. If it's not a 4xx response, like a 5xx response, it'll just return a simple error message with the status code